---
title: "A comprehensive tool suite for CNV analysis"
author:
  - name: Vinicius Henrique da Silva
    affiliation:
    - Animal Breeding and Genomics, Wageningen University & Research, Wageningen, The Netherlands
    - Animal Breeding and Genetics, Swedish University of Agricultural Sciences, Uppsala, Sweden
    - Netherlands Institute of Ecology (NIOO-KNAW), Wageningen, The Netherlands
    email: vinicius.dasilva@wur.nl
  - name: Ludwig Geistlinger
    affiliation: School of Public Health, City University of New York
    email: ludwig.geistlinger@sph.cuny.edu
package: cnvAnalyzeR
abstract: >
    The _cnvAnalyzeR_ package implements a comprehensive tool suite for CNV analysis.
    This includes functionality for summarizing individual CNV calls across a population,
    assessing overlap with functional genomic regions, and association analysis 
    with expression and quantitative phenotypes.
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{A comprehensive tool suite for CNV analysis}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r setup, echo=FALSE}
suppressPackageStartupMessages({ 
    library(cnvAnalyzeR)
    library(AnnotationHub)
    library(regioneR)
    library(BSgenome.Btaurus.UCSC.bosTau6.masked)
})
```

# Setup
Copy number variation (CNV) is a frequently observed deviation from the diploid 
state due to duplication or deletion of genomic regions.

CNVs can be experimentally detected based on comparative genomic hybridization [CGH]. 
On the other hand, CNVs can also be computationally inferred from SNP-arrays. 
A frequently used algorithm to infer CNVs from SNP-chip data is 
[PennCNV](http://penncnv.openbioinformatics.org), 
which shows good consistency with the CGH gold standard. 
In addition, recent developments have also enabled the detection of CNVs from 
next-generation sequencing data.

The named technologies for CNV detection have in common that they report, for each
sample under study, genomic regions that are duplicated or deleted with respect 
to a reference. 

Such regions are termed _CNV calls_ in the following and will be considered the 
starting point for analysis with the `r Biocpkg("cnvAnalyzeR")` package.  

The key parts of the functionality implemented in the package were developed,
described, and applied in several previous studies:

- Genome-wide detection of CNVs and their association with meat tenderness in 
    Nelore cattle
    [da Silva et al., 2016](https://doi.org/10.1371/journal.pone.0157711)

- Widespread modulation of gene expression by copy number variation in skeletal 
    muscle
    [Geistlinger et al., 2018](https://doi.org/10.1038/s41598-018-19782-4)

- CNVs are associated with genomic architecture in a songbird 
    [da Silva et al., 2018](https://doi.org/10.1186/s12864-018-4577-1)


# Reading and accessing CNV data

The `r Biocpkg("cnvAnalyzeR")` package uses Bioconductor core data strutures 
implemented in the `r Biocpkg("GenomicRanges")` and `r Biocpkg("RaggedExperiment")` 
packages to represent, access, and manipulate CNV data.

We start by loading the package.

```{r lib}
library(cnvAnalyzeR)
```

For demonstration, we consider CNV calls as obtained with 
[PennCNV](http://penncnv.openbioinformatics.org) from SNP-chip data in a 
Brazilian cattle breed. 
[da Silva et al., 2016](https://doi.org/10.1371/journal.pone.0157711)

```{r readCalls}
data.dir <- system.file("extdata", package="cnvAnalyzeR")
call.file <- file.path(data.dir, "Silva16_PONE_CNV_calls.csv")
calls <- read.csv(call.file, as.is=TRUE)
nrow(calls)
head(calls)
```

In total, there are roughly 50,000 calls obtained for 723 samples. 

```{r nrSamples}
length(unique(calls[,"NE_id"]))
```

We group the calls by sample ID, resulting in a `GRangesList`.
Each element of the list corresponds to a sample, and contains the genomic 
coordinates of the CNV calls for this sample (along with the copy number state 
in the `State` metadata column).  

```{r cnvCalls}
grl <- makeGRangesListFromDataFrame(calls, 
    split.field="NE_id", keep.extra.columns=TRUE)
grl
```

The advantage of representing the CNV calls as a `GRangesList` is that it allows
to leverage the comprehensive set of operations on genomic regions implemented 
in the `r Biocpkg("GenomicRanges")` packages - for instance, sorting of the calls 
according to their genomic coordinates.

```{r sortCalls}
grl <- sort(grl)
grl
```

An alternative matrix-like representation of the CNV calls can be obtained with 
the `r Biocpkg("RaggedExperiment")` data class. 
It resembles in many aspects the `r Biocpkg("SummarizedExperiment")`
data class for storing gene expression data as e.g. obtained with RNA-seq. 

```{r RaggedExperiment}
ra <- RaggedExperiment(grl)
ra
```

As apparent from the `dim` slot of the object, it stores the CNV calls in the 
rows and the samples in the columns. 
Note that the CN state is now represented as an assay matrix which can be 
easily accessed and subsetted.   

```{r RaggedExperiment-assay}
assay(ra[1:5,1:5])
```

As for a `r Biocpkg("SummarizedExperiment")`, additional information for 
the samples are annotated in the `colData` slot.
For example, we annotate the steer weight and its feed conversion ratio (FCR).
Feed conversion ratio is the ratio of dry matter intake to live-weight gain. 
A typical range of feed conversion ratios is 4.5 -7.5 with a lower number being 
more desirable as it would indicate that a steer required less feed per pound of
gain.

```{r RaggedExperiment-colData}
weight <- rnorm(ncol(ra), mean=1100, sd=100)
fcr <- rnorm(ncol(ra), mean=6, sd=1.5)
colData(ra)$weight <- round(weight, digits=2)
colData(ra)$fcr <- round(fcr, digits=2)
colData(ra)
```

# Summarizing individual CNV calls across a population

In CNV analysis, it is often of interest to summarize invidual calls across
the population, (i.e. to define CNV regions), for subsequent association
analysis with e.g. phenotype data.

In the simplest case, this just merges overlapping individual calls into 
summarized regions.

However, this typically inflates CNV region size and more appropriate approaches
have been developed for this purpose.

The `r Biocpkg("cnvAnalyzeR")` package implements three frequently used
approaches for summarizing CNV calls:

1. The [CNVRuler](http://www.ircgp.com/CNVRuler) procedure that trims region 
   margins based on regional density 
   [Kim et al., 2012](https://doi.org/10.1093/bioinformatics/bts239), 
2. the reciprocal overlap (RO) procedure that requires calls to sufficiently 
   overlap with one another 
   [Conrad et al., 2010](https://doi.org/10.1038/nature08516), and
3. the [GISTIC](http://www.broadinstitute.org/cancer/cga/gistic) procedure that
   identifies recurrent CNV regions 
   [Mermel et al., 2011](https://doi.org/10.1186/gb-2011-12-4-r41).


## Trimming low-density areas

Here, we use the approach from [CNVRuler](http://www.ircgp.com/CNVRuler) 
to summarize CNV calls to CNV regions (see 
[Figure 1](https://academic.oup.com/view-large/figure/83392426/bts239f1.jpeg)
in 
[Kim et al., 2012](https://doi.org/10.1093/bioinformatics/bts239)
an illustration of the approach). 

```{r cnvrs}
cnvrs <- populationRanges(grl, density=0.1)
cnvrs
```

## Reciprocal overlap

We also provide an implementation of the _Reciprocal Overlap (RO)_ procedure.


## Identifying recurrent regions

A `GISTIC`-like procedure for detecting recurrent CNVs is also in preparation.

# Overlap analysis of CNVs with functional genomic regions

Here, we compare the CNV overlap with distinct features in the genome using the `r Biocpkg("regioneR")` package.

Check for available _Bos taurus_ annotation from Ensembl: 
```{r getBtGenes}
library(AnnotationHub)
ah <- AnnotationHub()
ahDb <- query(ah, pattern = c("Bos taurus", "EnsDb"))
ahDb
```

Get genomic coordinates of genes for UMD3.1 assembly (Ensembl 92):
```{r getBtGenes2}
ahEdb <- ahDb[["AH60948"]]
bt.genes <- genes(ahEdb)
seqlevels(bt.genes) <- paste0("chr", seqlevels(bt.genes))
bt.genes
```

Select only genes on chromosome 1 and 2 for demonstration:
```{r formatBtGenes}
sel.genes <- bt.genes[seqnames(bt.genes) %in% c("1","2")]
sel.genes <- sel.genes[sel.genes$gene_biotype == "protein_coding"]
sel.genes
```

Perform overlap permutation test:
```{r ovlpTest}
library(regioneR)
library(BSgenome.Btaurus.UCSC.bosTau6.masked)
sel.cnvrs <- cnvrs[seqnames(cnvrs) %in% c("chr1", "chr2")]
res <- suppressWarnings(overlapPermTest(A=sel.cnvrs, B=sel.genes, ntimes=100, 
    genome="bosTau6", mask=NA, per.chromosome=TRUE, count.once=TRUE))
res
```

# CNV-expression association analysis

Here, we analyze differences in expression between CN states based on the
`r Biocpkg("edgeR")` package.


```{r rseqdata}
rcounts <- read.delim("RNAseq/all_counts_cleaned.txt", row.names=1, stringsAsFactors=FALSE)
rcounts <- as.matrix(rcounts)
dim(rcounts)
rcounts[1:5, 1:5]
```

```{r rse}
rse <- SummarizedExperiment(assays=list(rcounts=rcounts), 
                rowRanges=granges(bt.genes)[rownames(rcounts)])
rse
```


# CNV-phenotype association analysis

For CNV calls inferred from SNP-chip data, we additionally provide probe-by-probe
genome-wide association analysis (GWA) between CNVs and quantitative phenotypes as 
previously described ([da Silva et al., 2016](https://doi.org/10.1371/journal.pone.0157711)).

This treats common CN polymorphisms (CNPs, allele frequency >5%) as SNPs of 
equal frequency and carries out GWA as implemented in 
[PLINK](http://zzz.bwh.harvard.edu/plink/gvar.shtml).

For demonstration, we use CNV data of a wild population of songbirds 
([da Silva et al., 2018](https://doi.org/10.1186/s12864-018-4577-1)).

Using genomic estimated breeding values (GEBVs) for breeding time (BT) as the 
quantitative phenotype, we accordingly analyze for each CNV region whether change 
in copy number is associated with change in the genetic potential for breeding time. 

The function `setupCnvGWAS` imports phenotype information, CNV calls, and the probe map (if available).
The information required for analysis is then stored in the resulting `phen.info` list:

```{r importPhen}
## GEBV values
phen.loc <- file.path(data.dir, "Pheno.txt")
## CNV calls
cnv.out.loc <- file.path(data.dir, "CNVOut.txt")
## Genomic positions of the probes used in the CNV call
map.loc <-  file.path(data.dir, "MapPenn.txt")
phen.info <- setupCnvGWAS(name="example", 
    phen.loc=phen.loc, cnv.out.loc=cnv.out.loc, map.loc=map.loc)
phen.info 
```

The last item of the list displays the working directory:

```{r Wdir}
all.paths <- phen.info$all.paths
all.paths
```

For the GWAS, chromosome names are assumed to be `integer` (i.e. `1, 2, 3, ...`).
Non-integer chromosome names can be encoded by providing a `data.frame` that 
describes the mapping from `character` names to corresponding integers. 

For the example data, chromosomes _1A_, _4A_, _25LG1_, _25LG2_, and _LGE22_ 
are correspondingly encoded via 

```{r CNVGWASNames}
# Define chr correspondence to numeric
chr.code.name<- data.frame(   
                    V1=c(16, 25, 29:31), 
                    V2=c("1A", "4A", "25LG1", "25LG2", "LGE22"))
```

We can then run the actual CNV-GWAS, here without correction for multiple testing
which is *only done for exemplification*. 
In real analyses, multiple testing correction is recommended to avoid inflation 
of false positive findings.  

```{r CNVGWA}
segs.pvalue.gr <- cnvGWAS(phen.info, chr.code.name=chr.code.name, method.m.test="none")
segs.pvalue.gr
```

The CNV-GWAS uses the concept of CNV segments to define CNV loci. 
This procedure was originally proposed in our previous work in Nelore cattle 
([da Silva et al., 2016](https://doi.org/10.1371/journal.pone.0157711)) and 
defines CNV segments based on CNV genotype similarity of subsequent SNP probes. 

The default is `min.sim=0.95`, which will continuously add probe positions to a 
given CNV segment until the pairwise genotype similarity drops below 95%. 
Detailed up-down CNV genotype concordance that are used for the construction of 
CNV segments will look like the S12 Table in 
[da Silva et al., 2016](https://doi.org/10.1371/journal.pone.0157711). 

As `PLINK` returns a _p_-value for each probe, only one of the _p_-values of the
probes contained in a CNV segment is chosen as the segment _p_-value. 

This is similar to an approach used in differential expression (DE) analysis of 
microarray gene expression data, where typically the most significant DE probe
is chosen in case of multiple probes mapping to the same gene. 

Here, the representative probe for the CNV segment can be chosen to be the probe 
with lowest _p_-value (`assign.probe="min.pvalue"`, default) or the one with highest 
CNV frequency (`assign.probe="high.freq"`). 

Multiple testing correction based on the number of CNV segments tested is carried
out using the FDR approach (default). 
Results can then be displayed as for regular GWAS via a Manhattan plot 
(which can optionally be exported to a pdf file). 

```{r manh}
## Define the chromosome order in the plot
order.chrs <- c(1:24, "25LG1", "25LG2", 26:28, "LGE22", "1A", "4A")

## Chromosome sizes
chr.sizes <- c(114059860, 150265477, 111636321, 68030631, 61875929, 34342011,
                37653027, 31324166, 25106277, 20202851, 20315886, 20466350, 
                16480340, 16193477, 13820886, 10486032, 11572643, 9871655, 
                14661763, 7693166, 4276343, 6655392, 6808513, 1092960, 809223,
                6596997, 4327975, 5101010, 773534, 71365269, rep(19934923,2))

chr.size.order <- data.frame(chr=order.chrs, sizes=chr.sizes, stringsAsFactors=FALSE)

## Plot a pdf file with a manhatthan within the 'Results' workfolder
plotManhattan(all.paths, segs.pvalue.gr, chr.size.order, plot.pdf=FALSE)
```

CNV detection using SNP-chip intensities and allele frequencies is usually 
challenging and can produce biased CNV frequencies 
([da Silva et al., 2018](https://doi.org/10.1186/s12864-018-4577-1)). 

Therefore, we also provide the option to carry out the GWAS based on the 
relative signal intensity (log R ratio, LRR).
GWAS using LRR values directly facilitates the identification of CNV segments 
associated with the phenotype, disregarding putatively biased frequency estimates 
from the CNV calling procedure. 

To perform the GWAS using LRR values, import the LRR/BAF values and set `run.lrr=TRUE` 
in the `cnvGWAS` function:

```{r importLRR}
# List files to import LRR/BAF 
files <- list.files(data.dir, pattern = "\\.cnv.txt.adjusted$")
samples <- sub(".cnv.txt.adjusted$", "", files)
samples <- sub("^GT","", samples)
sample.files <- data.frame(file.names=files, sample.names=samples)
 
# All missing samples will have LRR = '0' and BAF = '0.5' in all SNPs listed in the GDS file
importLRR_BAF(all.paths, data.dir, sample.files)

# Read the GDS to check if the LRR/BAF nodes were added
cnv.gds <- file.path(all.paths[1], "CNV.gds")
(genofile <- SNPRelate::snpgdsOpen(cnv.gds, allow.fork=TRUE, readonly=FALSE))
gdsfmt::closefn.gds(genofile)

# Run the CNV-GWAS with existent GDS
segs.pvalue.gr <- cnvGWAS(phen.info, chr.code.name=chr.code.name, produce.gds=FALSE, run.lrr=TRUE)
```

It is important to note that the `cnvGWAS` function uses the `CNV.gds` file 
which is stored in the `Inputs` folder (i.e. `all.paths[1]`). 
This GDS file is automatically generated during a GWAS run. 

Therefore, running a GWAS implies that any GDS file produced by previous analysis 
will be overwritten. 
Use `produce.gds=FALSE` to avoid overwriting in the GWAS run. 

For convenience, a GDS file can be produced before the GWAS analysis with the 
`prodGdsCnv` function.
This additionally returns a `GRanges` object containing the genomic 
position, name and, frequency of each probe used to construct the CNV segments
for the GWAS analysis. 

Note that `probes.cnv.gr` object contains the integer chromosome names 
(as the GDS file on disk). 
Only the `segs.pvalue.gr`, which stores the GWAS results, has the character 
chromosome names.  

```{r prodGDS}
## Create a GDS file in disk and export the SNP probe positions
probes.cnv.gr <- prodGdsCnv(phen.info, chr.code.name=chr.code.name)
probes.cnv.gr

## Run GWAS with existent GDS file
segs.pvalue.gr <- cnvGWAS(phen.info, chr.code.name=chr.code.name, produce.gds=FALSE)
```


